Moving multi_inspiral utils to coherent module

a-r-williamson opened the pull request:
As development on `pycbc_multi_inspiral` has progressed, the executable has gathered a lot of utility functions to do with trigger identification and coherent SNR calculation that would be better placed elsewhere. I suggest moving them to a coherent module.

titodalcanton commented:
@a-r-williamson since you are introducing a new file, this is a good chance of applying the [Black autoformatter](https://github.com/gwastro/pycbc/wiki/How-to-satisfy-CodeClimate) without polluting the history of that file with code reformatting. Could you have a look?

a-r-williamson commented:
> @a-r-williamson since you are introducing a new file, this is a good chance of applying the [Black autoformatter](https://github.com/gwastro/pycbc/wiki/How-to-satisfy-CodeClimate) without polluting the history of that file with code reformatting. Could you have a look?

Having just spent some time going through trying to clean up a lot of the formatting by hand... I'll check that out, thanks!

titodalcanton commented on bin/pycbc_multi_inspiral:
```diff
@@ -16,56 +16,56 @@
-    description="Find multiple detector gravitational-wave triggers.")
+parser = argparse.ArgumentParser(
+    usage="", description="Find multiple detector gravitational-wave triggers."
+)
```
I suggest putting the description in the docstring at the top of the file, and then doing
```python
parser = argparse.ArgumentParser(description=__doc__)
```
See for example `pycbc_optimal_snr` or other scripts.

titodalcanton commented on bin/pycbc_multi_inspiral:
```diff
@@ -93,275 +93,64 @@ parser.add_argument("--user-tag", type=str, metavar="TAG", help="""
-# Use the time in the middle of the segment to calculate the antenna patterns.
-t_gps = opt.trigger_time
+log_level = logging.DEBUG if args.verbose else logging.INFO
+logging.basicConfig(format='%(asctime)s : %(message)s', level=log_level)
```
Use `pycbc.init_logging()` here.

a-r-williamson replied:
Done in 95caa1cc

titodalcanton commented on bin/pycbc_multi_inspiral:
```diff
@@ -567,167 +360,190 @@ with ctx:
-                    raise ValueError('The SNR triggers dictionary is empty.\
-                                     This should not be possible.')
+                    raise ValueError('The SNR triggers dictionary is empty. '
+                                     'This should not be possible.')
```
A `RuntimeError` may be more appropriate than a `ValueError` here.

titodalcanton commented on bin/pycbc_multi_inspiral:
```diff
@@ -736,25 +552,27 @@ with ctx:
 
 logging.info("Finished")
-logging.info("Time to complete analysis: %s" % str(time.time() - time_init))
+logging.info("Time to complete analysis: %d", int(time.time() - time_init))
```
This is a bit shorter and properly rounded:
```python
logging.info("Time to complete analysis: %.0f s", time.time() - time_init)
```

titodalcanton commented on pycbc/events/coherent.py:
```diff
@@ -0,0 +1,371 @@
+    # in at least 2 detectors.
+    coinc_idx = np.unique(coinc_list, return_counts=True)[0][
+        np.unique(coinc_list, return_counts=True)[1] > 1
+    ]
```
This may be more readable like this:
```python
something = np.unique(coinc_list, return_counts=True)
coinc_idx = something[0][something[1] > 1]
```

a-r-williamson replied:
Done in f222a2

titodalcanton commented on pycbc/events/coherent.py:
```diff
@@ -0,0 +1,371 @@
+
+
+def reweighted_snr(netwk_snr, netwk_chisq, idx1=3, idx2=1.0 / 6.0):
+    """Return the chi-squeared re-weighted SNR statistic
```
Typo.

a-r-williamson replied:
Done in f222a2

titodalcanton reviewed (changes requested):
@a-r-williamson this looks mostly ok, I have a few minor requests above, plus the following one: in the docstrings of the various reweighted SNRs and chi-squared functions, could you add links to the papers giving the definitions of those quantities?

a-r-williamson commented on pycbc/events/coherent.py:
```diff
@@ -0,0 +1,371 @@
+    """
+    denom = ((1 + netwk_chisq) ** idx1) / 2
+    rw_snr = netwk_snr / denom ** idx2
+    return rw_snr
```
This should be the same re-weighting. I'll switch to use `newsnr`

a-r-williamson commented:
@titodalcanton (and @pannarale) I think I've addressed all your comments, Tito. I'd love to get this merged as soon as possible, especially as this includes important fixes for bugs that we discovered while tidying up here:

[Serious bug in assigning SNRs and trigger array indices when looping over IFOs](https://github.com/gwastro/pycbc/commit/e532771e412563a51b8686227220c6aa75e62f50)
Each iteration in [this loop](https://github.com/gwastro/pycbc/commit/e532771e412563a51b8686227220c6aa75e62f50#diff-218a84325b9ee909bdcdc6b7e5d423c8b3e634b7176bb0c1a6b5514ea61b6e37L345-R362) was improperly assigning the outputs of `matched_filter_and_cluster`. These need to be copies. 

[Bug in the null SNR re-weighting](https://github.com/gwastro/pycbc/pull/3929/commits/16b6a7b35ff66d8fadcc5225f0cf3a9a1f399b66#diff-4b37abfdbeda70e2b7dbd03499c15377e8b4566d89e7e59523ad2bf0fccb48b9L347-R367)
The re-weighting of the detction statistic was being incorrectly applied with a threshold flat in null SNR. It should have been flat in null SNR up to a threshold in coherent SNR, then increase linearly with coherent SNR * some tunable gradient.

titodalcanton commented:
@a-r-williamson the diff for `pycbc_multi_inspiral` appears messed up somehow, but it may simply be extremely large. Could you maybe avoid the `if __name__ == '__main__':` block? Since this script is not supposed to be imported, I am not sure it adds anything, and it steals one indentation level across the *entire* script. I hope it was not `black` that did that :)

a-r-williamson commented:
@titodalcanton Done, thanks.

titodalcanton commented on bin/pycbc_multi_inspiral:
```diff
@@ -381,353 +165,359 @@ with ctx:
+    for t_num, template in enumerate(bank):
+        for ifo in args.instruments:
+            if args.cluster_method == "window":
+                cluster_window = int(args.cluster_window * sample_rate)
+            elif args.cluster_method == "template":
                 cluster_window = int(template.chirp_length * sample_rate)
-            elif opt.cluster_window == 0:
+            elif args.cluster_window == 0:
                  cluster_window = int(0)
```
Am I missing the point of having this if-else block inside the two for loops? It looks like you could easily move it to just before the loops.

titodalcanton replied:
Ah right, it does depend on the template, but can you at least move it before the loop over the instruments?

titodalcanton replied:
Actually, if you only use `cluster_window` in `cluster_template_network_events()`, then it is probably more readable to move this entire if-else block right before that call.

titodalcanton replied:
Also, `int(0)` is equivalent to just `0`.

a-r-williamson replied:
Well spotted. I'll move this down and remove the last check as it's redundant.

titodalcanton commented:
@a-r-williamson I have one last comment, and you should also fix codeclimate's points about redefining names. These are my last comments I think.
